### PR TITLE
allow adding lambda functions as an extension

### PR DESCRIPTION
**Pull request by soramichi**
This PR allows adding a lambda function as an extension,
resulting in simpler coding for simple extensions such as printing some values for debugging.

Example usage:

    ...
    trainer.extend(lambda x: print(model.accuracy.data), trigger=(1, "iteration"))
    ...

### CODE REVIEW

**Comment by muupan** [comment]
Thanks for your PR!

`isinstance(extension, types.FunctionType)` is not enough for checking whether it is a lambda. A function that is wrapped by `make_extension` also has `types.FunctionType`. It is possible that it has a non-`None` `finalize` attribute, but this PR would overwrite it with `None`.

Currently, extensions are expected to have `name` (or `default_name`) and `finalize` attributes, which is the reason it doesn't accept `lambda`. Instead of checking whether it is lambda or not, I suggest loosening the requirements by handling the case when it doesn't have such attributes:

- When `extension` doesn't have `name` nor `default_name`, then it is possible to check whether it has `__name__` and use it if yes, otherwise raise an error.
- `entry.extension.finalize` can be replaced with `getattr(entry.extension, 'finalize', None)`

By doing so, it can accept not only lambda but also any arbitrary callable object whose argument is `Trainer`.

In addition, this PR will affect the definition of `Extension` in Chainer, which is written in the docstring of `Extension` class: https://github.com/pfnet/chainer/blob/master/chainer/training/extension.py. Could you modify it accordingly?

**Comment by soramichi** [comment]
Thank you for the comments and suggestions.

I modified the code and the docstring in according to your comments. The docstring is bit confusing because if you add a lambda (or any callable that do **not** inherit Exntesion class), is is never fired unless you explicitly specify the `trigger` argument in the `extend` method.

It might be a good idea to change `get_trigger` function in `training/util.py` so that it returns (1, 'epoch') if `trigger` is None, but I didn't do it as it may break other constrains.

**Comment by muupan** [comment]
`Extension`'s docstring in `training/extension.py` contains more detailed explanation on how to define custom extensions. Could you modify it too?

As you wrote, modifying `get_trigger` would affect the behavior of existing code that explicitly set the `trigger` attribute to `None`. How about `getattr(extension, 'trigger', None)` -> `getattr(extension, 'trigger', (1, 'iteration'))` in https://github.com/pfnet/chainer/pull/1930/files#diff-768d99d739390d9a9a936299d6fa7fceR216 instead? It would not affect the existing code since extensions that inherit `Extension` or use `make_extension` must have the `trigger` attribute.

**Comment by soramichi** [comment]
Added fixes (docstring, trigger) as suggested and also added tests for lambda-based and callable class-based extensions.

**Comment by soramichi** [comment]
Thanks for the detailed review.

I changed the docstring of extension.py as suggested, and for trainer.py I wrote:
`If it is ``None`` and the extension does not have the trigger attribute`
instead of:
`If the extension does not have the trigger attribute`
because checking if an extension has the trigger attribute or not is done only when the `trigger` parameter given to `extend` is `None`.

```Python
        if trigger is None:
            trigger = getattr(extension, 'trigger', (1, 'iteration'))
        trigger = trigger_module.get_trigger(trigger)
```

File: chainer/training/trainer.py [hunk 1]
```diff
@@ -184,7 +185,10 @@ def extend(self, extension, name=None, trigger=None, priority=None,
             trigger (tuple or Trigger): Trigger object that determines when to
                 invoke the extension. If it is ``None``, ``extension.trigger``
                 is used instead. If the trigger is not callable, it is passed
-                to :class:`IntervalTrigger` to build an interval trigger.
+                to :class:`IntervalTrigger` to build an interval trigger. If
+                it is ``None`` and the extension does not have the trigger
+                attribute, the extension is triggered at every iteration as
```
Comment (delta2323) [inline chainer/training/trainer.py]: I think "by default" is better.

File: chainer/training/extension.py [hunk 2]
```diff
@@ -12,11 +12,16 @@ class Extension(object):
...
-    reduce the overhead and is much easier to use, while this class provides
-    more flexibility (for example, it can have methods to configure the
-    behavior).
+    There are three ways to define users' own extensions: inheriting this
+    class, decorating closures by :func:`make_extension`, or using any callable
+    including lambda functions as extensions. Decorator can slightly reduce the
+    overhead and is much easier to use, while this class provides more
+    flexibility (for example, it can have methods to configure the behavior).
+    Using a lambda function allows one-line coding for simple purposes, but
+    users have to specify the configurations in :meth:`Trainer.extend` by
```
Comment (delta2323) [inline chainer/training/extension.py]: What do you mean by "the configurations in Trainer.extend" ? Is it the argument of the method?

File: chainer/training/extension.py [hunk 3]
```diff
@@ -12,11 +12,16 @@ class Extension(object):
...
+    themselves. For a callable class not inheriting this class, the default
```
Comment (delta2323) [inline chainer/training/extension.py]: "a callable", not "a callable class" is better because there is a callable that is not a class (e.g. functions).

File: chainer/training/trainer.py [hunk 4]
```diff
@@ -39,10 +39,11 @@ class Trainer(object):
...
 
     Extensions are callable objects that take the trainer object as the
-    argument. There are two ways to define custom extensions: inheriting the
-    :class:`Extension` class, and decorating functions by
-    :func:`make_extension`. See :class:`Extension` for more details on custom
-    extensions.
+    argument. There are three ways to define custom extensions: inheriting the
+    :class:`Extension` class, decorating functions by :func:`make_extension`,
+    and defining any callable including lambda functions. The default
+    configurations of an extension is used in the former two cases. See
```
Comment (delta2323) [inline chainer/training/trainer.py]: For complete explanation, It is better to write what happens in the third case.

File: chainer/training/trainer.py [hunk 5]
```diff
@@ -184,7 +185,10 @@ def extend(self, extension, name=None, trigger=None, priority=None,
             trigger (tuple or Trigger): Trigger object that determines when to
                 invoke the extension. If it is ``None``, ``extension.trigger``
                 is used instead. If the trigger is not callable, it is passed
-                to :class:`IntervalTrigger` to build an interval trigger.
+                to :class:`IntervalTrigger` to build an interval trigger. If
```
Comment (delta2323) [inline chainer/training/trainer.py]: I think the sentence begin with "If it is None and the extension ... as default" should be placed that after the sentence "If it is None, extension.trigger is used instead." because it would make the document more readable.

File: tests/chainer_tests/training_tests/test_trainer.py [hunk 6]
```diff
@@ -56,13 +79,26 @@ def test_elapsed_time_serialization(self):
...
+        self.trainer.run()
+        self.assertTrue(dummy_callable_class.is_called)
+        self.assertTrue(dummy_callable_class.is_finalized)
+
+    def test_add_lambda_extension(self):
+        dummy_class = DummyClass()
+        self.trainer.extend(lambda x: dummy_class.touch())
+        self.trainer.run()
+        self.assertTrue(dummy_class.is_touched)
+
```
Comment (delta2323) [inline tests/chainer_tests/training_tests/test_trainer.py]: Could you test extending with a regular function just in case?

File: chainer/training/trainer.py (hunk 4 above)
Comment (muupan) [inline chainer/training/trainer.py]: I agree that it is not complete. How about removing the sentence `The default configurations of an extension is used in the former two cases.` because the complete explanation is already written in the class `Extension`?

Comment (soramichi) [inline chainer/training/trainer.py]: Thank you for the comments, but I don't fully understand what you mean.
Does "the third case" mean using any callable as an extension? I'm not sure what to explain for this case because there's really nothing that happens in this case (maybe we should explain the fact that nothing happens).

Comment (soramichi) [inline chainer/training/trainer.py]: Modified the docstring as follows. The idea is not to write too much here (as @muupan suggested)  but just mention that how to configuration is also in the docstring of `Extension`.
```
     Extensions are callable objects that take the trainer object as the
     argument. There are three ways to define custom extensions: inheriting the
     :class:`Extension` class, decorating functions by :func:`make_extension`,
     and defining any callable including lambda functions. See
     :class:`Extension` for more details on custom extensions and how to
     configure them.
```

File: chainer/training/trainer.py (hunk 5 above)
Comment (soramichi) [inline chainer/training/trainer.py]: Yes, that's what I also though a bit when I fixed originally. Changed the order as suggested.

File: chainer/training/trainer.py (hunk 1 above)
Comment (soramichi) [inline chainer/training/trainer.py]: Fixed as suggested.

File: chainer/training/extension.py (hunk 2 above)
Comment (soramichi) [inline chainer/training/extension.py]: Sorry for a hard-to-parse sentence. Fixed as follows:
```
users have to specify the configurations as arguments to :meth:`Trainer.extend`.
```
(should be "the" arguments? I'll fix again if needed)

File: chainer/training/extension.py (hunk 3 above)
Comment (soramichi) [inline chainer/training/extension.py]: Fixed as suggested.

File: tests/chainer_tests/training_tests/test_trainer.py (hunk 6 above)
Comment (soramichi) [inline tests/chainer_tests/training_tests/test_trainer.py]: Yes, I added a similar test for a regular function.

**Comment by soramichi** [comment]
@delta2323 @muupan Thank you for the very detailed check and suggestions. I added two commits and replied to all the reviews on by one, please check them out.

**Comment by muupan** [comment]
LGTM.

@delta2323 Can you review the changes?

**Comment by delta2323** [comment]
LGTM
